content/en/blog/2020/on-demand-egress-gateways/index.md

```diff
@@ -0,0 +1,540 @@
+---
+title: On Demand Egress Gateways
+subtitle: Deploy Egress Gateways independently in a seamless way
+description: Deploying Egress Gateways independently to fine-grained control of egress communication.
+publishdate: 2020-10-28
+attribution: Antonio Berben (At Deutsche Telekom - PAN-NET)
+keywords: [configuration,egress,gateway,external,service]
+target_release: 1.7.0
+---
+
+At [`Deutsche Telekom::PAN-NET`](https://pan-net.cloud/aboutus), we have embraced `Istio` as the umbrella to cover our services.
+
+Unfortunately, there are services which have not been migrated to `kubernetes` yet or, simply, cannot be. Therefore, those ones have to remain as external services.
+
+## Scenario
+
+If you are familiar with Istio, one of the offered ways to connect to external services is through [Egress Gateway](/docs/tasks/traffic-management/egress/egress-gateway/).
+
+However, if you want to satisfy the [Single-responsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle), you will need to deploy multiple and individual (1..N) `Egress Gateways` as the picture shows.
+
+{{< image width="75%" ratio="45.34%"
+    link="./on-demand-egress-gateway-overview.svg"
+    alt="Overview multiple Egress Gateways"
+    caption="Overview multiple Egress Gateways"
+    >}}
+
+With this model, one `Egress Gateway` is in charge of only one upstream service.
+
+Although the `IstioOperator` spec allows you to deploy multiple `Egress Gateways`, the manifest can become unmanageable.
+
+{{< text yaml >}}
+apiVersion: install.istio.io/v1alpha1
+kind: IstioOperator
+[...]
+spec:
+    egressGateways:
+    - name: egressgateway-1
+      enabled: true
+    - name: egressgateway-2
+      enabled: true
+    [egressgateway-3, egressgateway-4 ,...]
+    - name: egressgateway-N
+      enabled: true
+[...]
+{{< /text >}}
+
+As a benefit of decoupling Egress Getaways from IstioOperator manifest, you have enabled the possibility of setting up custom Liveness probes to have both services (gateway and upstream service) aligned.
+
+As well, you can apply the sidecar pattern to inject, for example, OPA into the pod to perform authorization with complex rules ([OPA envoy plugin](https://github.com/open-policy-agent/opa-envoy-plugin)).
+
+{{< image width="75%" ratio="45.34%"
+    link="./on-demand-egress-gateway-authz.svg"
+    alt="Authorization with OPA and `healthcheck` to external service"
+    caption="Authorization with OPA and `healthcheck` to external"
+    >}}
+
+At this point, you might be convinced that this is the right path to go.
+
+Now, the problem is that `Istio` does not offer a `CRD` for `Egress Gateways`. **Pity!**
```

Comment:
Isn't an `IstioOperator` with empty profile not that? https://preliminary.istio.io/latest/docs/setup/upgrade/gateways/#installation-with-istioctl

Reply:
Interesting! I did not try this yet. I will tomorrow.

But at first glance, I still miss the full control of the deployment object, I think. The possibility to manipulate the deployment to, for example, add sidecars (like OPA) is quite tempting, don't you think? :smile:

Reply:
I realize that Istio is a moving target, so hard to keep up with new things 😄 but there's a pretty new feature to allow you to customize k8s settings in a component: http://localhost:1313/latest/docs/setup/install/istioctl/#customize-kubernetes-settings

It's used in one of our egress gateway tasks (see step 3) here: https://preliminary.istio.io/latest/docs/tasks/traffic-management/egress/wildcard-egress-hosts/#setup-egress-gateway-with-sni-proxy

Reply:
Definitely Istio is moving fast (and it is great!). However, we can't even follow up on that pace. We moved from 1.5 to 1.6 to 1.7 and we decided to stop upgrading since we started spending more time on upgrades than on features :laughing: .

Regarding your links... it is great! That `overlays` is exactly what was needed. To be fair, it sounds a bit tricky since one is forced to not apply the `IstioOperator` manifest,  but to generate the final manifests and then `kubectl apply`. But, personally, I am glad to see that as part of the new version of Istio.

What is your recommendation regarding this post? Trying to adjust to `Istio 1.8`?

Appreciated your thoughts. Thanks!

Reply:
Hi Antonio,

Istio 1.8 is coming out in a week or so, and this post would probably come out after that.  Would it be possible to update your post for 1.8?  If not completely, at least to say that it will be made easier by these new features.

Thanks,
Craig (steering/blog guy)

Reply:
p.s. I hear you on the pace of upgrades

Reply:
@antonioberben I would recommend at least one more upgrade to 1.8 before pausing, since 1.8 has improvements that I think are very helpful for your use case. We've also finally updated and tested (https://github.com/istio/istio.io/blob/master/README.md#testing-document-content) all of our [egress tasks](https://preliminary.istio.io/latest/docs/tasks/traffic-management/egress/), so I think they will now provide much better guidance for you.

Reply:
Tested. Great improvements in 1.8.0! :muscle: great job.
Regarding the date time the article was written, I have just kept the original. 
Please, if you have any other thoughts, let me know. Thank you!

Reply:
It's best that the date be set to the day that the post goes live. So once it's ready to merge, I would request one last update (using the current date), and then we will approve/merge. But, if the post relies on 1.8 features, that should happen on a day after 1.8 is released, which should be this week.